### ISSUE TITLE
[ENH] Type hints for primitives and string arguments

### ISSUE THREAD

**Issue by TonyBagnall** [aeon-toolkit/aeon#1454]
### Describe the feature or idea you want to propose

we would like to increase the use of type hints, and a good first issue is to add them for the simplest case when arguments are primitives (int, float, bool) or strings. Pick a package, look through the code and add type hints. Just for primitives for now, more complex types need a bit of thought. Many have npt.ArrayLike types, but we have not decided on standardising on that, so leave these alone. Dont feel you have to do them all, this is a good first issue to get you started. 

### Describe your proposed solution

for example, the constructor for the anomaly detection algorithm STRAY has these type hints

```
    def __init__(
        self,
        alpha: float = 0.01,
        k: int = 10,
        knn_algorithm: str = "brute",
        p: float = 0.5,
        size_threshold: int = 50,
        outlier_tail: str = "max",
    ):
 ```
but the segmentation algorithm CLASP does not

```
def _is_trivial_match(candidate, change_points, n_timepoints, exclusion_radius=0.05):
    """Check if a candidate change point is in close proximity to other change points.
```

**Comment by nileenagp** [comment]
Hello, 

I would like to start on this issue. I am taking up "aeon.classification.distance_based._time_series_neighbors.py" script to start with.

**Comment by nileenagp** [comment]
@all-contributors please add @nileenagp  for Code

**Comment by nileenagp** [comment]
Hello, 
I have a doubt here.  If I need to use Type[keras.optimizers.Optimizer] in type hints for a parameter,  Is it allowed to import "keras" as per the project standards,  when it is not required anywhere else in the  script?

**Comment by MatthewMiddlehurst** [comment]
No, if the dependency is optional it should not be imported at the top level of a file. This would exclude it from specific type hints like that I imagine.

**Comment by nileenagp** [comment]
Thanks for the confirmation, I will leave that type hint for now.

**Comment by YashviMehta03** [comment]
Hello, is this issue still open to work on ?
I would like to work on it , could you please assign it.

Thank you!

**Comment by MatthewMiddlehurst** [comment]
Hi @YashviMehta03, this one is a bit too big for a single person really. I would post here selecting a part of the package you want to add type hints too and go from there.

**Comment by YashviMehta03** [comment]
Hello @MatthewMiddlehurst , yes right.
So, these are the modules where I could add type hints since they are missing :

1) all the .py files under the following directories :
aeon/classification/convolution_based
aeon/classification/feature_based

2) this particular file
aeon/classification/early_classification/_probability_threshold.py

I would like to work on it , could you assign it?

**Comment by MatthewMiddlehurst** [comment]
@YashviMehta03 feel free to open a PR

**Comment by MatthewMiddlehurst** [comment]
Hi not going to assign this since it's a large issue people are taking on incrementally but feel free to open a PR.

**Comment by gulshan-123** [comment]
Instead of `Union[np.ndarray, List[np.ndarray]]`, should we directly use `ArrayLike`. For example:
```
X_train : np.ndarray or list of np.ndarray
        Train data in a 2d or 3d ndarray or list of arrays.
```
here X_train can type hinted as Union, but I think using `ArrayLike` will be user friendly

Similarly, should I write `random_state: Optional[Union[int, np.random.RandomState, np.random.Generator]]=None`
as in the parameter `random_state : int, RandomState instance or None, default=None` is given. Or should we simply write `random_state: Optional[int]`. I think in this case we must write the expanded form for support of all random states.

**Comment by MatthewMiddlehurst** [comment]
We do not accept all arraylikes though is my issue with that, only specific formats. If a parameter accepts multiple types you should include them all. Fine to look at other packages and make suggestions based on how they do it if you wish.

**Comment by gulshan-123** [comment]
As per the docstring, it is accepting `np.ndarray` or list of `np.ndarray`.
So, I think adding `np.ndarray | list[np.ndarray]` will work.
Should I go forward and make the changes?

**Comment by an04shu** [comment]
Hi @TonyBagnall  @MatthewMiddlehurst, I'd love to contribute to this issue but I’m still getting familiar with the project. Could I start with a simple file, maybe one that hasn’t been worked on yet? Any suggestions on which module I should begin with? Thanks

**Comment by MatthewMiddlehurst** [comment]
No issue doing a singular file. I cant really give a suggestion, but try to avoid ones which have a PR open

**Comment by Ahmed-Zahran02** [comment]
Hi @MatthewMiddlehurst 
I would like to work on the base module for this issue?

**Comment by rc-ig** [comment]
Related to this issue, I've seen instances where the type stated in the docstring is not consistent with the usage:

https://github.com/aeon-toolkit/aeon/blob/ed98b2bb46dd9a7f3814fc82518b940583d90e9a/aeon/transformations/series/_clasp.py#L410
https://github.com/aeon-toolkit/aeon/blob/ed98b2bb46dd9a7f3814fc82518b940583d90e9a/aeon/transformations/series/_clasp.py#L449

What would the preference by in these cases?

**Comment by MatthewMiddlehurst** [comment]
That is just incorrect documentation. You can have multiple types (i.e. float or None), but i would just change that to float assuming there are no other types.

**Comment by bhakunisagar512-ui** [comment]
Hi, @MatthewMiddlehurst 
I’d like to add primitive and string type hints to  
`aeon/utility/validation/_dependencies.py`.

I’ll keep the PR limited to this single file.

**Comment by TUSHARDHOKRIYA** [comment]
@aeon-actions-bot assign @TUSHARDHOKRIYA

**Comment by SebastianSchmidl** [comment]
Hi, please do not assign this issue to yourself because it's a large issue people are taking on incrementally.

You are welcome to open a PR to address a certain package or class. Please post the scope of your changes in this issue so that others do not work on the same parts.

Please also avoid adding a `fixes #2737` to your PR; use `relates to #2737`.

**Comment by stephanielees** [comment]
Hello,
I noticed that all the `.py` files in the module `aeon/classification/deep_learning` don't have type hints yet.

Can I work on it?

**Comment by vickysharma-prog** [comment]
Hi @MatthewMiddlehurst, I reviewed aeon/segmentation/_clasp.py and found several function parameters with primitive defaults that currently lack type hints. I plan to add type hints only for simple built-in parameter types (int, float, bool, str) where unambiguous. The scope will be limited to _clasp.py with no logic changes. I’ll open a PR once ready. Thanks!

**Comment by Nandinisingh07** [comment]
Hii!
I was going through this issue and noticed that _clasp.py in the segmentation module still has some functions without type hints. 
I’d like to work on adding primitive type hints (int, float, bool, str) there and keep the changes limited to that file. 
Please let me know if that sounds  okay.

**Comment by Nandinisingh07** [comment]
Hello
I've been working on adding primitive type hints as part of this issue.
So far I've submitted:
- PR #3366: type hints for ClaSPTransformer
- PR #3367: type hints for LogTransformer

I'd like to continue this work. Are there specific files you'd like 
prioritized, or should I work through the transformations module 
systematically?

**Comment by TonyBagnall** [comment]
Im just closing this now as its too generic, more targetting issues are better
